### PR TITLE
Add "Profile" as steppingstone for testing --fakeroot

### PR DESCRIPTION
**Pull request by mem**
As part of testing the --fakeroot functionality, sometimes it's
necessary to run the same singularity command under multiple
"personalities" or "profiles". For example "singularity build" can be
executed with privilege or as "singularity build --fakeroot", and they
should both succeed. It's also expected that the command fails if run as
a regular user.

As part of end-to-end testing, all these personalities need to be
tested, and the "Profile" concept introduced here helps with that. The
RunSingularity driver needs to be extended to support this.

Most of this comes from Cedric; this is mostly a reorganization of that
code, with some documentation sprinkled into it. A lot of this is me
guessing and filling in gaps from #3997.

Signed-off-by: Marcelo E. Magallon <marcelo@sylabs.io>

Attn: @singularity-maintainers

### CODE REVIEW

**Comment by gvallee** [comment]
Related to #4126

File: e2e/internal/e2e/priv_linux.go
Comment (sashayakovtseva) [inline e2e/internal/e2e/priv_linux.go]: Period.

File: e2e/internal/e2e/priv_linux.go
Comment (sashayakovtseva) [inline e2e/internal/e2e/priv_linux.go]: Period.

File: e2e/internal/e2e/profile/profile.go
Comment (sashayakovtseva) [inline e2e/internal/e2e/profile/profile.go]: Shouldn't you use `e2e.OrigGID()` here?

File: e2e/internal/e2e/profile/profile.go
Comment (sashayakovtseva) [inline e2e/internal/e2e/profile/profile.go]: Overall this looks okay, but I think what you are trying to do is to check whether one slice contains at least one element from the other.
Maybe this func should be changed to smth like `Contains(s, sub []string)`? and then in `Args()` you will do e.g :

```
if Contains() {
return []string{"--fakeroot"}
}
```

File: e2e/internal/e2e/profile/profile.go
Comment (sashayakovtseva) [inline e2e/internal/e2e/profile/profile.go]:
```suggestion
// RootUserNamespace implements Profile describing a privileged user that
```

File: e2e/internal/e2e/profile/profile.go
Comment (sashayakovtseva) [inline e2e/internal/e2e/profile/profile.go:204]: Honestly, I don't really like this double embedding. It took some time to determine why `RootUserNamespace .Privileged()` will refer `Root.Privileged()`. 
This is diamond embedding. I understand, that this may be convenient, but really hard to understand at the first glance.

**Review by sashayakovtseva — CHANGES_REQUESTED** [review]
A couple suggestions I want to discuss.

**Review by jscook2345 — DISMISSED** [review]
Approve pending discussion of @sashayakovtseva notes. Thanks!

File: e2e/internal/e2e/profile/profile.go
Comment (mem) [reply e2e/internal/e2e/profile/profile.go]: I had the exact same question! I'm glad I'm not alone :-)

The file that this is reading, the one specified by the first argument, is indexed by user name, not group name. I remember my original feedback when this function was added was that it should be two separate functions, not one function doing two different things, but some implementation details make it so that having a single function is more convenient.

I guess I can add a comment clarifying. Thanks.

Comment (jscook2345) [reply e2e/internal/e2e/profile/profile.go]: I see now what you mean after your description. Thanks.

File: e2e/internal/e2e/profile/profile.go
Comment (mem) [reply e2e/internal/e2e/profile/profile.go]: I think I see what you mean.

This function started off as something different, so I loss some perspective in the process.

This is taking something that already exists as a slice (cmd), which will in most cases contain a single element, but in some cases it will contain two, and maybe more.

Then it takes the list of commands, commands, and looks for cmd in that list. Since cmd is a slice and commands is a slice (not a slice of slices), what the code does is concatenate the elements of cmd and looks for _that_ in the commands slice. If it finds it, it then returns the flag argument a a slice.

All of this is super weird. I'll try the suggested change.

File: e2e/internal/e2e/profile/profile.go
Comment (mem) [reply e2e/internal/e2e/profile/profile.go:204]: Yeah, I had a different version of this with a single embedded struct. Then I switched to embedding the other. And then I ended up embedding both because that's what RootUserNamespace is: one part UserNamespace, one part Root, with Root defining the user and UserNamespace defining flags and requirements. The "purely imperative" version of this (for lack of a better way to describe it) is worse, because it has different switch cases depending on the function you are looking at.

The subtlety here is that these structs do not have overlapping methods. If both structs where directly defining the same method, say Privileged, you wouldn't be able to call RootUserNamespace.Privileged() without specifying which one you want (or without implementing Privileged directly for RootUserNamespace). That's Go's solution to the diamond problem: do not allow the ambiguity at the call location.

So, while I agree with you, I'm not sure what to do about it...

**Review by sashayakovtseva — APPROVED** [review]

**Comment by gvallee** [comment]
It would have been great to actually discuss this before merging. I have no idea how to use this, there is no example, no explanation, issue #4126 has *not* been updated and what @cclerget did is now totally irrelevant. @jscook2345 @GodloveD What is a path forward?

**Pull request merged**
